### ISSUE TITLE
Can't install with system PMI if system also has PMIx but not libevent

### ISSUE THREAD

**Issue by Flamefire** [open-mpi/ompi#9105]
## Background information

I want to reinstall OpenMPI the same as I did before but in the meantime the pmix-devel libs have been installed on the cluster. Now the OpenMPI configure process insists on using that and fails due to missing libevent-devel libs but I want to use neither!

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

Tested many version from 4.x

## Details of the problem

Configure aborts with 
```
configure: WARNING: EXTERNAL PMIX SUPPORT REQUIRES USE OF EXTERNAL LIBEVENT
configure: WARNING: LIBRARY. THIS LIBRARY MUST POINT TO THE SAME ONE USED
configure: WARNING: TO BUILD PMIX OR ELSE UNPREDICTABLE BEHAVIOR MAY RESULT
configure: error: PLEASE CORRECT THE CONFIGURE COMMAND LINE AND REBUILD
```

Cmdline is `./configure --with-slurm --with-pmi=/usr --with-pmi-libdir=/usr/lib64 --with-pmix=internal --with-knem=/opt/knem-1.1.3.90mlnx1 --enable-mpirun-prefix-by-default  --enable-shared  --with-hwloc=/opt/hwloc/2.0.3-GCCcore-8.3.0  --with-verbs`

Checking the m4 sources shows, that unless `--with-pmi=no` is given it will always force-use the system pmix which is NOT what I requested when doing `--with-pmix=internal`

**Comment by ggouaillardet** [comment]
Thanks for the report, there is indeed something wrong with the pmix detection.

can you give the inline patch below a try?
note you will need recent autotools, and run `./autogen.pl` (or `./autogen.pl --force` if you apply the patch on an official tarball)



```patch
diff --git a/config/opal_check_pmi.m4 b/config/opal_check_pmi.m4
index 087eba8..960d876 100644
--- a/config/opal_check_pmi.m4
+++ b/config/opal_check_pmi.m4
@@ -211,10 +211,11 @@ AC_DEFUN([OPAL_CHECK_PMI],[
 
            # check for pmix lib installed by slurm */
            slurm_pmix_found=no
-           OPAL_CHECK_PMIX_LIB([$check_pmi_install_dir],
-                               [$check_pmi_lib_dir],
-                               [opal_enable_pmix=yes],
-                               [opal_enable_pmix=no])
+           AS_IF([test "$opal_external_pmix_happy" = "yes" ],
+                 [OPAL_CHECK_PMIX_LIB([$check_pmi_install_dir],
+                                      [$check_pmi_lib_dir],
+                                      [opal_enable_pmix=yes],
+                                      [opal_enable_pmix=no])])
 
            # since support was explicitly requested, then we should error out
            # if we didn't find the required support
```

**Comment by ggouaillardet** [comment]
@rhc54 I think I spotted an other issue (not the root cause here)

```
checking if user requested PMI support... yes
checking for pmi.h in /usr/include... not found
checking for pmi.h in /usr/include/slurm... found
checking pmi.h usability... yes
checking pmi.h presence... yes
checking for pmi.h... yes
checking for PMI_Init in -lpmi... yes
checking for pmi2.h in /usr/include... not found
checking for pmi2.h in /usr/include/slurm... found
checking pmi2.h usability... yes
checking pmi2.h presence... yes
checking for pmi2.h... yes
checking for PMI2_Init in -lpmi2... yes
checking for pmix.h in ... not found
checking for pmix.h in /include... not found
checking can PMI support be built... yes
```

Open MPI was configure'd with `--with-pmi`
SLURM `pmi1` and `pmi2` were found in `/usr/include/slurm`, but SLURM's `pmix.h` header was searched in `/`
(I guess it should have been searched in `/usr/include/slurm` too).

**Comment by Flamefire** [comment]
I don't think the proposed patch would work. `opal_external_pmix_happy` is set by `OPAL_CHECK_PMIX_LIB` so from what I can tell it cannot be set at this point which basically disables this code.

Also related to this: `./configure --with-slurm --with-pmi=/usr
 --with-pmi-libdir=/usr/lib64 --with-knem=/opt/knem-1.1.3.90mlnx1 --enable-mpirun-prefix-by-default  --enable-shared  --with-hwloc=/opt/hwloc  --with-libevent=/opt/libevent  --with-ofi=/opt/libfabric  --with-pmix=/opt/PMIx  --with-ucx=/opt/UCX  --without-verb` runs into an even worse issue:
```
checking for pmix.h in /usr... not found
checking for pmix.h in /usr/include... found
checking libpmix.* in /usr/lib64... found
checking PMIx version... version file found
checking version 4x... not found
checking version 3x or above... found
checking can PMI support be built... yes
checking PMIx version to be used... external(3x)
```

Again it fully ignores what the user specified.
I don't understand at all why the pmi check codepath checks for pmix. That check should happen later, shouldn't it?

**Comment by ggouaillardet** [comment]
yep, `test "$with_pmix" != "internal"` should do the trick.

I did not author these bits, but I would guess the intention was to look for `SLURM` pmix headers in the ad-hoc `/usr/include/slurm` directory and/or treat SLURM `PMIx` just like `PMI1` and `PMI2`.

**Comment by Flamefire** [comment]
> yep, test "$with_pmix" != "internal" should do the trick.

This won't work for the use case above: `--with-pmix=/opt/PMIx`

I'd recommend removing that whole call. If the user wants to use the slurm PMIx he can use `--with-pmix=/usr/include/slurm`

**Comment by rhc54** [comment]
> @rhc54 I think I spotted an other issue (not the root cause here)
> 
> ```
> checking if user requested PMI support... yes
> checking for pmi.h in /usr/include... not found
> checking for pmi.h in /usr/include/slurm... found
> checking pmi.h usability... yes
> checking pmi.h presence... yes
> checking for pmi.h... yes
> checking for PMI_Init in -lpmi... yes
> checking for pmi2.h in /usr/include... not found
> checking for pmi2.h in /usr/include/slurm... found
> checking pmi2.h usability... yes
> checking pmi2.h presence... yes
> checking for pmi2.h... yes
> checking for PMI2_Init in -lpmi2... yes
> checking for pmix.h in ... not found
> checking for pmix.h in /include... not found
> checking can PMI support be built... yes
> ```
> 
> Open MPI was configure'd with `--with-pmi`
> SLURM `pmi1` and `pmi2` were found in `/usr/include/slurm`, but SLURM's `pmix.h` header was searched in `/`
> (I guess it should have been searched in `/usr/include/slurm` too).

Hi @ggouaillardet - afraid that has nothing to do with me. This is all OMPI "glue" code, so somebody on the OMPI side of things will need to figure it out 😄

**Comment by rhc54** [comment]
The snippet from Gilles is coming from the OMPI configure, not PMIx - OMPI is looking for the pmix.h header and not using the Slurm location.

**Comment by Flamefire** [comment]
@rhc54 You authored https://github.com/open-mpi/ompi/commit/0322ad028da3e586b22d771e08cf4d4230ff5507 which introduced the bug.
Can you double check your claim:
> If the user also specifies --with-pmix and gives a different location, then override anything previously found and look for it where the user directed.

Because this is (seemingly) not what happens. `OPAL_CHECK_PMIX_LIB` in the --with-pmi path sets `opal_enable_pmix=yes`
And the --with-pmix path is only entered when `AS_IF([test "$opal_enable_pmix" = "no"],`
I think this is a C&P bug on your side because you set `slurm_pmix_found=no` in that commit but never use that ever.
I see that this commit has merge conflict markers, so maybe something went wrong?

Would you consider my proposal from above:
> I'd recommend removing that whole call. If the user wants to use the slurm PMIx he can use --with-pmix=/usr/include/slurm

**Comment by ggouaillardet** [comment]
@Flamefire I think Ralph's comment was referring to his involvement in the Open MPI project.

Anyway, I am fine with the rationale described in the commit message you pointed, so as far as I am concerned,
the next move is to fix the bug when both `--with-pmi` and `--with-pmix` are passed

> If the user also specifies --with-pmix and gives a different location, then override anything previously found and look for it where the user directed.

**Comment by Flamefire** [comment]
> @Flamefire I think Ralph's comment was referring to his involvement in the Open MPI project.

My comment was unrelated to his response, just updating with information I found. Sorry for the confusion


> @Flamefire I think Ralph's comment was referring to his involvement in the Open MPI project.

I'd argue against that. We have our HPC clusters SLURM build with PMI2 and PMIx, so the PMIx headers are still there. However the libevent headers are not. We want to build OpenMPI with PMI2 support (only) and hence pass `--with-pmi`, given that `--with-pmix` exists IMO it is completely unexpected that it will build with PMIx too and, even worse, it fails because it did found only PMIx but not libevent and hence aborts.

So bottom line: Why is the semantic of `--with-pmi` different to `--with-pmix`?
The former (used to) mean, to build with pmi (1&2) if passed and without it if not. The latter by itself means the same: Build with pmix when passed and without if not. But now --with-pmi may or may not imply building with pmix which introduces actual breakages due to the implicit dependency on libevent. Furthermore (I assume) it does not allow to enable only PMI anymore as e.g. `--without-pmix`/`--with-pmix=off` (whatever it was) will be ignored as well.

**Comment by rhc54** [comment]
> @Flamefire I think Ralph's comment was referring to his involvement in the Open MPI project.

Yeah, one of the few benefits of getting older is retirement 😄 

FWIW: the thinking for many years was that OMPI should only be compiled with _one_ PMI component. It's kinda weird to pass more than that - you have three options (PMI-1, PMI-2, and PMIx). All provide the same basic capabilities so far as MPI is concerned, so if we look at implementation and feature coverage, they roughly (and arguably) would equate to "good, better, best".

The thought was that you always take the best available, and it makes no sense to use something less. So you would never pass both `--with-pmi` and `--with-pmix` as that makes no sense. If the system supports pmix, then use it - if not, then use pmi. The change I provided was to deal with the situation where a sys admin wanted to use the same configure script to cover both situations. Rare, but we thought understandable.

Your corner-case adds another dimension due to the libevent issue. You have the problem that OMPI itself requires libevent, so I'm not sure how you intend to get around that one. The libevent requirement is pretty fundamental to OMPI and cannot be avoided. So regardless of PMI choice, your build is going to have a libevent dependency.

Again, it's a new corner-case that was never considered. Can't help with the solution, I'm afraid - eventually, it boils down to limitations on the various corner-cases one can support as they inevitably begin to conflict with each other.

**Comment by Flamefire** [comment]
> So you would never pass both --with-pmi and --with-pmix as that makes no sense. If the system supports pmix, then use it - if not, then use pmi. The change I provided was to deal with the situation where a sys admin wanted to use the same configure script to cover both situations. Rare, but we thought understandable.

We do because we have a large, heterogeneous cluster with a common software installation and some parts support pmix and some only PMI(2). And even worse: Users who do `srun --mpi=pmi2` and refuse to change that.

> The change I provided was to deal with the situation where a sys admin wanted to use the same configure script to cover both situations. 

Can you elaborate on that? What does one stop from doing `--with-pmi --with-pmix`? That already covers that, doesn't it?

> You have the problem that OMPI itself requires libevent, so I'm not sure how you intend to get around that one. 

OMPI also requires PMIx and it enforces that you use either external PMIx *and* libevent or the internal (if no option was passed). But due to the change it will now run into a situation where it will autodetect a system PMIx although only system PMI was requested and insists to use that which then fails due to missing libevent.

**Comment by ggouaillardet** [comment]
There are several issues here

1. `--with-pmix=internal` was explicitly requested, but this is not what was selected by Open MPI. This is a bug and it will be fixed.
2. Using an external `PMIx` implicitly requires the header files of the external `libevent` so Open MPI can be built. This is probably something we should do a better job at documenting.
3. Open MPI could automatically fallback to the internal `PMIx` if no external `libevent` can be used. That could be nice to have, PR is welcome.

Regardless, my advice is you install the `libevent` headers and `*.so` symlinks (either ask your sysadmin, or do it yourself and `configure CPPFLAGS=-I$HOME/mylibevent/include LDFLAGS=-L$HOME/mylibevent/lib` so you can use SLURM PMIx and move forward.

**Comment by Flamefire** [comment]
For your 1.: Any value for `--with-pmix` is ignored, so even specifying another path.

And IMO the point that --with-pmi implies external pmix even though --with-pmix is not given is at least unexpected. But that just as feedback from a user/system maintainer.

FWIW: For our installation we patch out that code part in the --with-pmi path: https://github.com/easybuilders/easybuild-easyconfigs/pull/13307/files#diff-3690abc8347dbb87f8c6689872f472607b1ca778dcd7afbb8c3d606c9afe4eda

**Comment by rhc54** [comment]
>     2\. Using an external `PMIx` implicitly requires the header files of the external `libevent` so Open MPI can be built. This is probably something we should do a better job at documenting.

Just FYI: an external pmix mandates that you also use an external libevent and hwloc so we can ensure that the pmix and OMPI libraries are linked against the _same_ installations of those packages. This is actually checked in newer versions of OMPI during configure and will error out if the condition is not met. Probably should port that code to this release (if it isn't already there) and ensure this gets wired into it.

**Comment by Flamefire** [comment]
> This is actually checked in newer versions of OMPI during configure and will error out if the condition is not met

It is in all versions I tested and actually one of the 2 issues: Using an external PMI implies using an external PMIx (if it happens to be on the system) and the mentioned check makes configure fail

**Comment by ggouaillardet** [comment]
@Flamefire 

can you please give the inline patch below a try?
SLURM PMIx won't be tested if `--with-pmix=internal` or `--with-pmix=<path to PMIx>` is used, so I expect it to fix the bug you reported. Note in your case you still need to manually force `--with-pmix=internal` (since SLURM `PMIx` is present and an external `libevent` cannot be used - because the devel package is not installed)

```patch
diff --git a/config/opal_check_pmi.m4 b/config/opal_check_pmi.m4
index 087eba8..ca6002e 100644
--- a/config/opal_check_pmi.m4
+++ b/config/opal_check_pmi.m4
@@ -14,7 +14,7 @@
 # Copyright (c) 2011-2014 Los Alamos National Security, LLC. All rights
 #                         reserved.
 # Copyright (c) 2014-2019 Intel, Inc.  All rights reserved.
-# Copyright (c) 2014-2018 Research Organization for Information Science
+# Copyright (c) 2014-2021 Research Organization for Information Science
 #                         and Technology (RIST).  All rights reserved.
 # Copyright (c) 2016-2021 IBM Corporation.  All rights reserved.
 # $COPYRIGHT$
@@ -139,15 +139,6 @@ AC_DEFUN([OPAL_CHECK_PMI_LIB],
 AC_DEFUN([OPAL_CHECK_PMI],[
     OPAL_VAR_SCOPE_PUSH([check_pmi_install_dir check_pmi_lib_dir default_pmi_libloc slurm_pmi_found])
 
-    AC_ARG_WITH([pmi],
-                [AC_HELP_STRING([--with-pmi(=DIR)],
-                                [Build PMI support, optionally adding DIR to the search path (default: no)])],
-                                [], with_pmi=no)
-
-    AC_ARG_WITH([pmi-libdir],
-                [AC_HELP_STRING([--with-pmi-libdir=DIR],
-                                [Look for libpmi or libpmi2 in the given directory DIR, DIR/lib or DIR/lib64])])
-
     check_pmi_install_dir=
     check_pmi_lib_dir=
     default_pmi_libloc=
@@ -209,12 +200,12 @@ AC_DEFUN([OPAL_CHECK_PMI],[
                          opal_pmi2_rpath="$pmi2_rpath"
                          AC_SUBST(opal_pmi2_rpath)])])
 
-           # check for pmix lib installed by slurm */
-           slurm_pmix_found=no
-           OPAL_CHECK_PMIX_LIB([$check_pmi_install_dir],
-                               [$check_pmi_lib_dir],
-                               [opal_enable_pmix=yes],
-                               [opal_enable_pmix=no])
+           # check for pmix lib installed by slurm unless an other PMIx library is requested
+           AS_IF([test -z "$with_pmix" || test "$with_pmix" = "yes" || test "$with_pmix" = "external"],
+                 [OPAL_CHECK_PMIX_LIB([$check_pmi_install_dir],
+                                      [$check_pmi_lib_dir],
+                                      [opal_enable_pmix=yes],
+                                      [opal_enable_pmix=no])])
 
            # since support was explicitly requested, then we should error out
            # if we didn't find the required support
@@ -398,14 +389,6 @@ AC_DEFUN([OPAL_CHECK_PMIX_LIB],[
 
 AC_DEFUN([OPAL_CHECK_PMIX],[
 
-    AC_ARG_WITH([pmix],
-                [AC_HELP_STRING([--with-pmix(=DIR)],
-                                [Build PMIx support.  DIR can take one of three values: "internal", "external", or a valid directory name.  "internal" (or no DIR value) forces Open MPI to use its internal copy of PMIx.  "external" forces Open MPI to use an external installation of PMIx.  Supplying a valid directory name also forces Open MPI to use an external installation of PMIx, and adds DIR/include, DIR/lib, and DIR/lib64 to the search path for headers and libraries. Note that Open MPI does not support --without-pmix.])])
-
-    AC_ARG_WITH([pmix-libdir],
-                [AC_HELP_STRING([--with-pmix-libdir=DIR],
-                                [Look for libpmix the given directory DIR, DIR/lib or DIR/lib64])])
-
     AS_IF([test "$with_pmix" = "no"],
           [AC_MSG_WARN([Open MPI requires PMIx support. It can be built])
            AC_MSG_WARN([with either its own internal copy of PMIx, or with])
@@ -475,3 +458,30 @@ AC_DEFUN([OPAL_CHECK_PMIX],[
                  [OPAL_SUMMARY_ADD([[Miscellaneous]],[[PMIx support]], [opal_pmix], [External ($opal_external_pmix_version)])])],
           [OPAL_SUMMARY_ADD([[Miscellaneous]], [[PMIx support]], [opal_pmix], [Internal])])
 ])
+
+# Check both PMI and PMIx
+#
+# OPAL_CHECK_ALL_PMI()
+# --------------------------------------------------------
+AC_DEFUN([OPAL_CHECK_ALL_PMI],
+[
+    AC_ARG_WITH([pmi],
+                [AC_HELP_STRING([--with-pmi(=DIR)],
+                                [Build PMI support, optionally adding DIR to the search path (default: no)])],
+                                [], with_pmi=no)
+
+    AC_ARG_WITH([pmi-libdir],
+                [AC_HELP_STRING([--with-pmi-libdir=DIR],
+                                [Look for libpmi or libpmi2 in the given directory DIR, DIR/lib or DIR/lib64])])
+
+    AC_ARG_WITH([pmix],
+                [AC_HELP_STRING([--with-pmix(=DIR)],
+                                [Build PMIx support.  DIR can take one of three values: "internal", "external", or a valid directory name.  "internal" (or no DIR value) forces Open MPI to use its internal copy of PMIx.  "external" forces Open MPI to use an external installation of PMIx.  Supplying a valid directory name also forces Open MPI to use an external installation of PMIx, and adds DIR/include, DIR/lib, and DIR/lib64 to the search path for headers and libraries. Note that Open MPI does not support --without-pmix.])])
+
+    AC_ARG_WITH([pmix-libdir],
+                [AC_HELP_STRING([--with-pmix-libdir=DIR],
+                                [Look for libpmix the given directory DIR, DIR/lib or DIR/lib64])])
+
+    OPAL_CHECK_PMI
+    OPAL_CHECK_PMIX
+])
diff --git a/configure.ac b/configure.ac
index 62ed652..01d055e 100644
--- a/configure.ac
+++ b/configure.ac
@@ -20,8 +20,8 @@
 # Copyright (c) 2013      Mellanox Technologies, Inc.
 #                         All rights reserved.
 # Copyright (c) 2013-2019 Intel, Inc.  All rights reserved.
-# Copyright (c) 2014-2017 Research Organization for Information Science
-#                         and Technology (RIST). All rights reserved.
+# Copyright (c) 2014-2021 Research Organization for Information Science
+#                         and Technology (RIST).  All rights reserved.
 # Copyright (c) 2016-2017 IBM Corporation.  All rights reserved.
 # Copyright (c) 2018      Amazon.com, Inc. or its affiliates.
 #                         All Rights reserved.
@@ -299,8 +299,7 @@ m4_ifdef([project_oshmem],
 OPAL_CONFIGURE_OPTIONS
 OPAL_CHECK_OS_FLAVORS
 OPAL_CHECK_CUDA
-OPAL_CHECK_PMI
-OPAL_CHECK_PMIX
+OPAL_CHECK_ALL_PMI
 m4_ifdef([project_orte], [ORTE_CONFIGURE_OPTIONS])
 m4_ifdef([project_ompi], [OMPI_CONFIGURE_OPTIONS])
 m4_ifdef([project_oshmem], [OSHMEM_CONFIGURE_OPTIONS])

```

**Comment by Flamefire** [comment]
Yes that patch works for me

**Comment by rhc54** [comment]
@Flamefire Just as a "heads-up": starting with OMPI v5.0 (out this summer), OMPI will no longer support PMI-1 or PMI-2. Only PMIx. Just FYI so you can start planning.

**Comment by ggouaillardet** [comment]
@rhc54 yep, the patch has been merged and there was no complain about it.
I am now closing this issue.
